lib/spack/llnl/util/argparsewriter.py

```diff
@@ -10,195 +10,363 @@
+            prog (str): the program name
+            out (file object): the file to write to
+            aliases (bool): whether or not to include subparsers for aliases
+        """
+        super(ArgparseWriter, self).__init__(prog)
         self.level = 0
+        self.prog = prog
         self.out = out
-
-    def _write(self, parser, root=True, level=0):
+        self.aliases = aliases
+
+    def parse(self, parser, prog):
+        """Parses the parser object and returns the relavent components.
+
+        Parameters:
+            parser (argparse.ArgumentParser): the parser
+            prog (str): the command name
+
+        Returns:
+            str: the command name
+            str: the command description
+            str: the command usage
+            list: list of positional arguments
+            list: list of optional arguments
+            list: list of subcommand parsers
+        """
         self.parser = parser
-        self.level = level
-        actions = parser._actions
 
-        # allow root level to be flattened with rest of commands
-        if type(root) == int:
-            self.level = root
-            root = True
+        split_prog = parser.prog.split(' ')
+        split_prog[-1] = prog
+        prog = ' '.join(split_prog)
+        description = parser.description
+
+        fmt = parser._get_formatter()
+        actions = parser._actions
+        groups = parser._mutually_exclusive_groups
+        usage = fmt._format_usage(None, actions, groups, '').strip()
 
-        # go through actions and split them into optionals, positionals,
+        # Go through actions and split them into optionals, positionals,
         # and subcommands
         optionals = []
         positionals = []
         subcommands = []
         for action in actions:
             if action.option_strings:
-                optionals.append(action)
+                flags = action.option_strings
+                dest_flags = fmt._format_action_invocation(action)
+                help = self._expand_help(action) if action.help else ''
+                help = help.replace('\n', ' ')
+                optionals.append((flags, dest_flags, help))
             elif isinstance(action, argparse._SubParsersAction):
                 for subaction in action._choices_actions:
                     subparser = action._name_parser_map[subaction.dest]
-                    subcommands.append(subparser)
+                    subcommands.append((subparser, subaction.dest))
+
+                    # Look for aliases of the form 'name (alias, ...)'
+                    if self.aliases:
+                        match = re.match(r'(.*) \((.*)\)', subaction.metavar)
+                        if match:
+                            aliases = match.group(2).split(', ')
+                            for alias in aliases:
+                                subparser = action._name_parser_map[alias]
+                                subcommands.append((subparser, alias))
```

Comment:
It took me an entire day to figure out how to include aliases of Spack commands and this is what I came up with. It's still not perfect, as it can lead to duplication. For example, if we added an alias like `spack comp -> spack compiler`, it would add all of the `spack compiler *` subsubcommands twice instead of `spack comp *` subcommands. Luckily this situation doesn't currently exist.